Aliu/fabric init infra

aliuTT opened the pull request:
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/16728)

### Problem description
Fabric bringup testing is more or less stable now. We want to launch fabric automatically on scaleout systems, and we need to move the kernels/configs to Metal runtime.

### What's changed
- Moved `tt_fabric` to be part of `Metalium` library. This is needed since things like device pool and dispatch now has dependency. There was also an ask from Artem/Bryan Wilder to possibly reduce libraries under `tt-metal`. 

- Add fabric config to `device/topology.cpp`
- Added apis to launch Fabric.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

aliuTT committed:
Move fabric to tt_metal

aliuTT committed:
Integrate fabric init infra to metal runtime

aliuTT commented on tt_metal/api/tt-metalium/tt_metal.hpp:
```diff
@@ -23,11 +23,16 @@ class IDevice;
 inline bool InMainThread() { return not InWorkerThread(); }
 
+// Call before CreateDevices to enable fabric, which uses all ethernet cores and some tensix cores
+void InitializeFabricSetting(detail::FabricSetting fabric_setting);
```
I will move this into the CreateDevices call.

aliuTT replied:
To be honest, I'm 50/50 on having this as a separate API. I think I will leave it as is then

nhuang-tt commented on tt_metal/impl/device/device_pool.cpp:
```diff
@@ -451,6 +509,31 @@ void DevicePool::init_firmware_on_active_devices() const {
+        std::filesystem::path(tt::llrt::RunTimeOptions::get_instance().get_root_dir()) /
+        "tt_metal/fabric/mesh_graph_descriptors" / mesh_graph_descriptor;
+
+    std::cout << "initialize control plane with mesh graph descriptor: " << mesh_graph_desc_path << std::endl;
```
Replace with logger or remove?`std::cout`

aliuTT replied:
will remove

nhuang-tt reviewed (commented):
Is there a way to disable it if we don't want to enable fabric

TT-BrianLiu commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -1626,30 +1644,37 @@ int main(int argc, char **argv) {
             traffic.notify_tx_controller();
         }
 
+        std::cout << "done notify tx workers" << std::endl;
```
Remove

aliuTT replied:
will remove

TT-BrianLiu commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -1626,30 +1644,37 @@ int main(int argc, char **argv) {
             tt_metal::detail::WaitProgramDone(test_device->device_handle, test_device->program_handle);
         }
-
+        std::cout << "done wait for program done" << std::endl;
```
Remove

aliuTT replied:
will remove

aliuTT commented on tt_metal/api/tt-metalium/dispatch_core_manager.hpp:
```diff
@@ -119,6 +120,20 @@ class dispatch_core_manager {
         chip_id_t upstream_device_id, chip_id_t device_id, uint16_t channel, uint8_t cq_id);
 
     const tt_cxy_pair& us_tunneler_core_local(chip_id_t device_id, uint16_t channel, uint8_t cq_id);
+    /// @brief Gets the location of the kernel desginated for fabric gatekeeper.
+    /// @param device_id ID of the device that a fast dispatch command targets
+    /// @return tt_cxy_pair logical location (chip + core coordinate) of the fabric_gatekeeper core
```
no

aliuTT commented on tt_metal/api/tt-metalium/dispatch_core_manager.hpp:
```diff
@@ -119,6 +120,20 @@ class dispatch_core_manager {
+        return assignment.fabric_gatekeeper.value();
+    }
 
     /// @brief Gets the location of the kernel desginated to write to the completion queue region for a particular command queue
     ///         Each command queue has one completion queue
```
no

aliuTT commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -1626,30 +1644,37 @@ int main(int argc, char **argv) {
+                test_device->wait_for_router_sync();
+            }
         }
+        std::cout << "done wait for rrouter sync" << std::endl;
```
will remove

aliuTT commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -1626,30 +1644,37 @@ int main(int argc, char **argv) {
+            }
         }
 
+        std::cout << "done terminate router kernel" << std::endl;
```
will remove

aagarwalTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -773,58 +762,39 @@ std::unique_ptr<Program> create_and_compile_fabric_program(IDevice* device) {
     constexpr uint32_t default_test_results_addr = 0x100000;
     constexpr uint32_t default_test_results_size = 0x40000;
```
are these used anywhere? I think we should be able to get rid of these

aagarwalTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -752,17 +752,6 @@ void configure_dispatch_cores(IDevice* device) {
-
 std::unique_ptr<Program> create_and_compile_fabric_program(IDevice* device) {
     auto fabric_program_ptr = std::make_unique<Program>();
     constexpr uint32_t default_tunneler_queue_size_bytes = 0x8000;  // maximum queue (power of 2)
```
shall we rename these to avoid any potential confusion?

aagarwalTT commented on tt_metal/impl/device/device_pool.cpp:
```diff
@@ -642,17 +639,14 @@ void DevicePool::close_devices(const std::vector<IDevice*>& devices) {
 
     // Terminate fabric gatekeeper
```
nit: update comment to avoid confusion

aagarwalTT commented on tt_metal/llrt/tt_cluster.cpp:
```diff
@@ -145,12 +146,32 @@ void Cluster::generate_cluster_descriptor() {
+        if (all_n300) {
+            if (this->cluster_desc_->get_all_chips().size() == 1) {
+                this->cluster_type_ = ClusterType::N300;
+            } else if (this->cluster_desc_->get_all_chips().size() == 8) {
```
maybe this is handled somewhere else, but can we just rely on the number of chips to determine if its TG? We should also be checking for the required connections b/w the n300 boards?

aagarwalTT commented on tt_metal/impl/device/device_pool.cpp:
```diff
@@ -373,12 +390,53 @@ void DevicePool::add_devices_to_pool(const std::vector<chip_id_t>& device_ids) {
+void DevicePool::wait_for_fabric_master_router_sync() const {
+    if (this->fabric_setting == detail::FabricSetting::FABRIC) {
+        std::vector<uint32_t> master_router_terminate(1, 0);
+        auto fabric_router_sync_sem_addr =
```
Should we use the `FABRIC_ROUTER_SYNC_SEM ` from `tt_fabric_interface.h` to keep things consistent?

aliuTT replied:
`tt_fabric_interface.h` has arch dependencies, I cannot include it in host code. I have a static assert somewhere to assert that the hal value and `FABRIC_ROUTER_SYNC_SEM` are the same

tt-dma commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -14,370 +15,371 @@
-    {0, 0, 0, 0, PREFETCH_HD, {x, x, x, x}, {1, 2, x, x}, NOC::NOC_0, NOC::NOC_0, NOC::NOC_0},
-    {1, 0, 0, 0, DISPATCH_HD, {0, x, x, x}, {2, x, x, x}, NOC::NOC_0, NOC::NOC_1, NOC::NOC_0},
-    {2, 0, 0, 0, DISPATCH_S, {0, x, x, x}, {1, x, x, x}, NOC::NOC_1, NOC::NOC_1, NOC::NOC_1},
+    {0, 0, 0, 0, PREFETCH_HD, {X, X, X, X}, {1, 2, X, X}, NOC::NOC_0, NOC::NOC_0, NOC::NOC_0},
```
Is this auto-formatting? Wondering if we can make an exception here as it makes the tables much harder to read.

aliuTT replied:
I'll undo these changes, I didn't need the .x after cleanup

aagarwalTT commented on tt_metal/api/tt-metalium/dispatch_core_manager.hpp:
```diff
@@ -119,6 +120,20 @@ class dispatch_core_manager {
+    /// @brief Gets the location of the kernel desginated for fabric gatekeeper.
+    /// @param device_id ID of the device that a fast dispatch command targets
+    /// @return tt_cxy_pair logical location (chip + core coordinate) of the fabric_gatekeeper core
+    const tt_cxy_pair& fabric_gatekeeper(chip_id_t device_id) {
+        // One fabric gatekeeper per device, no dependency on channel or cq_id
+        dispatch_core_placement_t& assignment = this->fabric_core_assignments[device_id];
+        if (assignment.fabric_gatekeeper.has_value()) {
+            return assignment.fabric_gatekeeper.value();
+        }
+        CoreCoord gatekeeper_coord = this->get_next_available_dispatch_core(device_id);
+        assignment.fabric_gatekeeper = tt_cxy_pair(device_id, gatekeeper_coord.x, gatekeeper_coord.y);
+        log_dispatch_assignment("Fabric Gatekeeper", assignment.fabric_gatekeeper.value(), device_id, 0, 0);
+        return assignment.fabric_gatekeeper.value();
+    }
```
we can get rid of this now?

aagarwalTT commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -1626,30 +1644,37 @@ int main(int argc, char **argv) {
-        // wait for all routers to handshake with master router
-        for (auto& [chip_id, test_device] : test_devices) {
-            test_device->wait_for_router_sync();
+        if (metal_fabric_init_level == 0) {
```
Should we also have this check for creating router kernels from the test code? 

`        for (auto& [chip_id, test_device] : test_devices) {
            test_device->create_router_kernels(router_compile_args, defines);
        }`

aliuTT replied:
I have added this locally.

aagarwalTT commented on tests/tt_metal/tt_metal/perf_microbenchmark/routing/test_tt_fabric_sanity.cpp:
```diff
@@ -85,11 +86,15 @@ typedef struct test_board {
+    std::unique_ptr<tt::tt_fabric::ControlPlane> cp_owning_ptr;
     uint32_t num_chips_to_use;
     std::string mesh_graph_descriptor;
+    uint32_t metal_fabric_init_level;
```
I think we can keep this global

aagarwalTT commented on tt_metal/api/tt-metalium/dispatch_core_manager.hpp:
```diff
@@ -40,6 +40,7 @@ struct dispatch_core_placement_t {
     std::optional<tt_cxy_pair> mux_d = std::nullopt; // Mux
     std::optional<tt_cxy_pair> demux_d = std::nullopt; // Demux
     std::optional<tt_cxy_pair> tunneler_d = std::nullopt; // ethernet tunneler
+    std::optional<tt_cxy_pair> fabric_gatekeeper = std::nullopt;  // Fabric Gatekeeper
```
remove?

aagarwalTT commented on tt_metal/api/tt-metalium/dispatch_core_manager.hpp:
```diff
@@ -190,6 +205,7 @@ class dispatch_core_manager {
     // {device ID : {channel (hugepage) : {cq_id : dispatch assignment}}}
     // Each device has an assigned hugepage at a specific channel that holds (up to 2) hardware command queues (represented by cq_id)
     std::unordered_map<chip_id_t, std::unordered_map<uint16_t, std::unordered_map<uint8_t, dispatch_core_placement_t>>> dispatch_core_assignments;
+    std::unordered_map<chip_id_t, dispatch_core_placement_t> fabric_core_assignments;
```
I think this can be removed as well, now that we are not launching gatekeeper?

tt-dma commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -750,4 +752,73 @@ void configure_dispatch_cores(IDevice* device) {
     }
 }
 
+std::unique_ptr<Program> create_and_compile_fabric_program(IDevice* device) {
```
Should we make Graphs/Nodes for the fabric kernels? Now that we have a new way of setting up dispatch kernels I think it makes sense to stick to it for fabric kernels as well

aliuTT replied:
let's do this in the next step when we hook up FD to use fabric

aagarwalTT commented on tt_metal/api/tt-metalium/fabric_host_interface.h:
```diff
@@ -0,0 +1,65 @@
+// TODO: move routing table here
+namespace tt::tt_fabric {
+
+constexpr uint32_t GATEKEEPER_INFO_SIZE_BYTES = 848;
+constexpr uint32_t GATEKEEPER_INFO_SYNC_OFFSET = 0;
```
Is there a way to remove this hardcoding? Or maybe we can get rid of this completely now that gatekeeper is gone

aliuTT replied:
I've removed

aagarwalTT commented on tt_metal/impl/dispatch/dispatch_core_manager.cpp:
```diff
@@ -293,6 +293,7 @@ dispatch_core_manager::dispatch_core_manager(const DispatchCoreConfig& dispatch_
 void dispatch_core_manager::reset_dispatch_core_manager(
     const DispatchCoreConfig& dispatch_core_config, uint8_t num_hw_cqs) {
     this->dispatch_core_assignments.clear();
+    this->fabric_core_assignments.clear();
```
not needed with gk gone?

aagarwalTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -750,4 +752,73 @@ void configure_dispatch_cores(IDevice* device) {
+
+    for (const auto& router_logical_core : device->get_active_ethernet_cores()) {
+        // initialize the semaphore
+        auto fabric_router_sync_sem_addr =
```
Should we use the `FABRIC_ROUTER_SYNC_SEM`  from `tt_fabric_interface.h` to keep things consistent?

tt-dma commented on tt_metal/impl/dispatch/topology.hpp:
What's the plan for merging fabric/fd topologies? Will fabric kernels just replace mux/demux/router/tunneler, and be specified in the dispatch topology structs?

aliuTT replied:
Plan is murky right now. We are going to support both 1D Fabric and 2D Fabric, so we may need to keep the mux/demux/router/tunneler kernels to support current FD tunneling on N300/T3K. We will also have a config where FD directly talks to Fabric, and the mux/demux/router/tunneler kernels will not be there.

tt-dma commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -750,4 +752,73 @@ void configure_dispatch_cores(IDevice* device) {
+    // TODO: the size below is overriding eth barrier, need to fix this
+    constexpr uint32_t default_tunneler_test_results_size = 0x6000;  // 256kB total L1 in ethernet core - 0x39000
+    constexpr uint32_t default_test_results_addr = 0x100000;
+    constexpr uint32_t default_test_results_size = 0x40000;
```
Hard-coded addresses to either dispatch_constants (fabric_constants?) or through the hal?

nhuang-tt replied:
I think for the prod environment like in here, test results args should all be 0/disabled. And in the kernels, those counters related to the test results need to be compiled out.
I disabled all this stuff in the mux/demux kernels and it can actually increases the B/cycle.

aliuTT replied:
I'll remove these

omilyutin-tt commented on tt_metal/fabric/hw/inc/tt_fabric_interface.h:
This file has:
`#define INVALID 0x0`
During the fabric hackathon last week, this failed to compile TTNN because it collided with 
`enum class Layout { ROW_MAJOR = 0, TILE = 1, INVALID = 2 };`

Can we get rid of all macros and use `constexpr uint32_t`?

aliuTT replied:
This if file movement, not part of this PR. Can you open an issue for this to track? In any case, this is a hw only include, not sure why it's included in host code.

tt-aho replied:
This was done as a hack for the hackathon. Should not base review/decisions based on compile issues from the hackathon branch since it was never intended to be merged.
This file is currently included in dedicated metal test host code, which is compiled per arch which is why it was okay. Was never expected to be a host level production include without review/rework if needed.

omilyutin-tt commented on tt_metal/distributed/mesh_device.cpp:
```diff
@@ -635,6 +635,10 @@ void MeshDevice::init_command_queue_device() {
     TT_THROW("init_command_queue_device() is not supported on MeshDevice - use individual devices instead");
     reference_device()->init_command_queue_device();
 }
+void MeshDevice::init_fabric() {
```
What is the plan for this?

omilyutin-tt reviewed (changes requested):
Also https://github.com/tenstorrent/tt-metal/issues/16725 :)

Many suggestions were non-blocking, some clarifications / suggestions, but some of the critical ones that I think we actually need to fix:
* Minimize the number of macros -- see one of the comments below about "INVALID" breaking TTNN. There is no reason to use macros over `constexpr` for variables.
* The interfaces for routing tables are hard to use. Unwrap `vector<vector<vector>>>` into strong type accessors.
* Related: `mesh_id_t` (and all similar IDs) need to be strong types, and not just aliases to `int`. Right now is the easiest time to do it. Also, let's not do `_t` in C++.
* Tangling fabric with `DevicePool` needs to be clarified. This singleton is already problematic, we are creating more problems for ourselves by adding more state management logic to it.

omilyutin-tt commented on tt_metal/impl/device/device_pool.cpp:
```diff
@@ -277,6 +289,11 @@ void DevicePool::initialize_device(IDevice* dev) const {
 
+    // TODO: add handling of EDM
+    if (this->fabric_setting == detail::FabricSetting::FABRIC) {
+        dev->init_fabric();
```
Why here, as opposed to after `init_command_queue_device`?

aliuTT replied:
Because CQ/FD is a client of Fabric.

omilyutin-tt commented on tt_metal/impl/device/device_pool.cpp:
```diff
@@ -373,12 +390,53 @@ void DevicePool::add_devices_to_pool(const std::vector<chip_id_t>& device_ids) {
             this->activate_device(device_id);
         }
     }
+    // Only can launch Fabric if all devices are active
+    if (this->fabric_setting == detail::FabricSetting::FABRIC) {
+        for (int i = 0; i < tt::Cluster::instance().number_of_devices(); i++) {
+            if (not _inst->is_device_active(i)) {
+                // Fabric currently requires all devices to be active
+                log_warning(tt::LogMetal, "Fabric is disabled because device {} is not active", i);
+                this->fabric_setting = detail::FabricSetting::DISABLED;
+                break;
+            }
+        }
+    }
+
+    // TODO: add handling of EDM
+    if (this->fabric_setting == detail::FabricSetting::FABRIC) {
+        // Initialize control plane, which writes routing tables to all ethernet cores
+        _inst->initialize_control_plane();
```
This is hard to track:
`DevicePool::initialize()` calls `add_devices_to_pool()`, which also calls `activate_device()` for each of the active device. There, we silently disable fabric if not all devices are active.

`DevicePool::initialize()` then calls `init_firmware_on_active_devices()` that iterates over active devices, and initializes them. There we hit a check that if fabric isn't active, we shouldn't initialize it.

Can we TT_FATAL if the user tries to use fabric but not all devices are active? Also what is the difference between "DEFAULT" and "DISABLED"?

aliuTT replied:
I'm intending DEFAULT to launch "best" mode Metal picks. DISABLED will explicitly not launch any fabric kernels.

omilyutin-tt replied:
Can you rename to something like "AUTO" and explain in the enum documentation?

aliuTT commented on tt_metal/api/tt-metalium/tt_metal.hpp:
```diff
@@ -23,11 +23,16 @@ class IDevice;
 
 namespace detail {
 
+enum class FabricSetting { DISABLED = 0, FABRIC = 1, EDM = 2, DEFAULT = 3 };
```
It can be called Config

aliuTT commented on tt_metal/api/tt-metalium/mesh_graph.hpp:
```diff
@@ -57,39 +57,36 @@ using InterMeshConnectivity = std::vector<std::vector<std::unordered_map<mesh_id
 using IntraMeshConnectivity = std::vector<std::vector<std::unordered_map<chip_id_t, RouterEdge>>>;
```
Cleaning this up is not part of this PR.

aliuTT commented on tt_metal/api/tt-metalium/fabric_host_interface.h:
```diff
@@ -0,0 +1,65 @@
+    (sizeof(std::uint32_t) / sizeof(chan_id_t)) == NUM_CHANNELS_PER_UINT32,
+    "LOG_BASE_2_NUM_CHANNELS_PER_UINT32 must be equal to log2(sizeof(std::uint32_t) / sizeof(chan_id_t))");
+
+enum eth_chan_magic_values {
```
File movement, not part of this PR. Can you open an issue for this to track?

aliuTT commented:
@omilyutin-tt addressing the comments you listed as critical explicitly, please review and unblock if possible:
- Minimize the number of macros -- see one of the comments below about "INVALID" breaking TTNN. There is no reason to use macros over constexpr for variables.
  - These are hw includes, we can make an effort to clean up all these files, but my changes here are only to move the file.
- The interfaces for routing tables are hard to use. Unwrap vector<vector<vector>>> into strong type accessors.
Related: mesh_id_t (and all similar IDs) need to be strong types, and not just aliases to int. Right now is the easiest time to do it. Also, let's not do _t in C++.
    - Valid. Tracked by https://github.com/tenstorrent/tt-metal/issues/16725, we need to schedule in the cleanup, but not going to be this PR.

- Tangling fabric with DevicePool needs to be clarified. This singleton is already problematic, we are creating more problems for ourselves by adding more state management logic to it.
    - I also found it difficult to find the correct place to put Fabric init. It must be initialized before fast dispatch, and after device launches base FW. I see no other place to put fabric init since both of these are handled in DevicePool. To disentangle fabric with DevicePool, we need to remove fast dispatch init from DevicePool as well.

@aagarwalTT thanks for all the suggestions! I will cleanup the legacy gatekeeper objects.

aliuTT committed:
Add auto fabric init tests to CI

tt-aho commented on tt_metal/fabric/control_plane.cpp:
```diff
@@ -489,6 +489,7 @@ void ControlPlane::write_routing_tables_to_chip(mesh_id_t mesh_id, chip_id_t chi
             tt::Cluster::instance().l1_barrier(physical_chip_id);
         }
     }
+    tt::Cluster::instance().l1_barrier(physical_chip_id);
```
Should the barrier on line 489 be removed since we have the barrier here now?

aliuTT replied:
yea, I didn't notice this due to no merge conflicts. Not sure why the author of line 492 put a barrier per ethernet channel.

ubcheema commented:
Do we need to update the slow dispatch tests for fabric to leverage the new fabric launch mechanism?

aliuTT commented:
> Do we need to update the slow dispatch tests for fabric to leverage the new fabric launch mechanism?

Yea, I added the `--metal_fabric_init_level` toggle to the benchmarks for this purpose.
